src/Services/FileLoader/CommonMarkExtensions/HeadingWrapper.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+
+namespace App\Services\FileLoader\CommonMarkExtensions;
+
+
+use League\CommonMark\Block\Element\AbstractBlock;
+use League\CommonMark\Block\Element\Heading;
+use League\CommonMark\Block\Renderer\BlockRendererInterface;
+use League\CommonMark\ElementRendererInterface;
+use League\CommonMark\HtmlElement;
+
+class HeadingWrapper implements BlockRendererInterface
+{
+    private $headers = [];
+
+    public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, $inTightList = false)
+    {
+        $render = $htmlRenderer->renderInlines($block->children());
+        $pureTitle = strip_tags($render);
+
+        if ($block instanceof Heading) {
+            $this->headers[] = ["text" => $pureTitle, "level" => $level = $block->getLevel(), "id" => $id = "h$level-" . $this->slugify($pureTitle) . "-" . substr(md5($render), 0, 10)];
+        }
+        return new HtmlElement('h' . $level, ["id" => $id], [$render]);
+    }
+
+
+
+    public function extractTOC()
+    {
+
+        $tree = [];
+        $lastOfLevel = [];
+        foreach ($this->headers as &$c) {
+            if (!isset($lastOfLevel[$c["level"]])) {
+                if (count($lastOfLevel) > 0) {
+                    if (min(array_keys($lastOfLevel)) > $c["level"]) {
+                        $lastOfLevel = [];
+                    }
+                }
+
+                if (count($lastOfLevel) < 1) {
+                    $tree[] = &$c;
+                }
+                $lastOfLevel[$c["level"]] = &$c;
+                for ($i = $c["level"] - 1; $i > 0; $i--) {
+                    if (isset($lastOfLevel[$i])) {
+                        if (!isset($lastOfLevel[$i]["children"])) {
+                            $lastOfLevel[$i]["children"] = [];
+                        }
+                        $lastOfLevel[$i]["children"][] = &$c;
+                        $lastOfLevel[$c["level"]] = &$c;
+                        break;
+                    }
+                }
+
+            } else {
+
+                for ($i = $c["level"] - 1; $i > 0; $i--) {
+                    if (isset($lastOfLevel[$i])) {
+                        $lastOfLevel[$i]["children"][] = &$c;
+                        break;
+                    }
+                }
+                for ($i = $c["level"]; $i < 6; $i++) {
+                    unset($lastOfLevel[$i]);
+                }
+                if (count($lastOfLevel) < 1) {
+                    $tree[] = &$c;
+                }
+                $lastOfLevel[$c["level"]] = &$c;
+
+            }
+        }
+        return $tree;
+
+    }
+
+    private function slugify($string, $delimiter = '-')
```

Comment:
Quick comment to summarize what will be cleaned after this function. Or a link to a standard "slugify" process explanation